x-pack/plugins/infra/public/plugin.ts

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+import { merge } from 'lodash';
+import {
+  Plugin as PluginClass,
+  CoreSetup,
+  CoreStart,
+  PluginInitializerContext,
+  AppMountParameters,
+} from 'kibana/public';
+import { InMemoryCache, IntrospectionFragmentMatcher } from 'apollo-cache-inmemory';
+import ApolloClient from 'apollo-client';
+import { ApolloLink } from 'apollo-link';
+import { HttpLink } from 'apollo-link-http';
+import { withClientState } from 'apollo-link-state';
+import { DEFAULT_APP_CATEGORIES } from '../../../../src/core/utils';
+import { InfraFrontendLibs } from './lib/lib';
+import introspectionQueryResultData from './graphql/introspection.json';
+import { InfraKibanaObservableApiAdapter } from './lib/adapters/observable_api/kibana_observable_api';
+import { registerStartSingleton } from './legacy_singletons';
+import { registerFeatures } from './register_feature';
+import { HomePublicPluginSetup, HomePublicPluginStart } from '../../../../src/plugins/home/public';
+import { DataPublicPluginSetup, DataPublicPluginStart } from '../../../../src/plugins/data/public';
+import { UsageCollectionSetup } from '../../../../src/plugins/usage_collection/public';
+import { DataEnhancedSetup, DataEnhancedStart } from '../../data_enhanced/public';
+import { LogsRouter, MetricsRouter } from './routers';
+
+export type ClientSetup = void;
+export type ClientStart = void;
+
+export interface ClientPluginsSetup {
+  home: HomePublicPluginSetup;
+  data: DataPublicPluginSetup;
+  usageCollection: UsageCollectionSetup;
+  dataEnhanced: DataEnhancedSetup;
+}
+
+export interface ClientPluginsStart {
+  home: HomePublicPluginStart;
+  data: DataPublicPluginStart;
+  dataEnhanced: DataEnhancedStart;
+}
+
+export type InfraPlugins = ClientPluginsSetup & ClientPluginsStart;
+
+const getMergedPlugins = (setup: ClientPluginsSetup, start: ClientPluginsStart): InfraPlugins => {
```

Comment:
Plugins should almost never have to mix setup and start API's. When calling `await core.getStartServices();` it looks a bit like both start and setup api's are used at the same time, but this will only resolve once setup is complete. Once setup is complete, these API's should no longer be used since it means that lifecycle has finished. 

Using `InfraPlugins = ClientPluginsStart;` doesn't generate any typescript errors, so it's probably safe to remove `getMergedPlugins` and just use pluginsStart directly.

Reply:
> Plugins should almost never have to mix setup and start API's.

Yep, makes complete sense. I originally had things setup the way you've suggested (using only `ClientPluginsStart`), but I ran into issues with the `usageCollection` plugin. It's `undefined` in `start`, but I have access to things in `setup`. I did ask about this on Slack at some point, and it was pointed out that there's no `start` contract, only a `setup` one, so I assumed it was okay to do this.

We've tried to make all our access to these through the `useKibana` hook, so I need a way at the top level to still get access to `reportUiStats`.

In case this helps:

<img width="1210" alt="Screenshot 2020-02-14 13 35 08" src="https://user-images.githubusercontent.com/471693/74536037-2588c900-4f2f-11ea-8afa-ec1d9ce12e2d.png">

Reply:
Yeah if a plugin doesn't expose it's API from start, then what you've done is the only way to work around that. I'll follow-up with Ahmad, but I think `reportUiStats` belongs in `start`. However, you don't seem to be using this API anywhere?

Reply:
Cool, thank you 👍 We're not using it anywhere in the `infra` codebase directly, but we're using the shared `useTrackPageview` hook from the `observability` plugin, which uses `reportUiStats`. We haven't really documented it anywhere yet but we were working with the idea each observability plugin would need to setup it's own `<KibanaContextProvider />` to make use of said hooks.